datalake: add redpanda.iceberg.invalid.record.action topic property

nvartolomei opened the pull request:
This PR only adds the property. It is not being used yet. Working in parallel on using it. Happy to merge as-is or the larger PR which is upcoming.

Please pay extra attention to this PR as I have never added a topic property before.

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v24.3.x
- [ ] v24.2.x
- [ ] v24.1.x

## Release Notes

* none

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

nvartolomei committed:
config: introduce iceberg_invalid_record_action

nvartolomei committed:
cluster: iceberg_invalid_record_action topic property

nvartolomei committed:
rptest: improve assert messages in test

nvartolomei commented:
/dt

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/61092-->
<!--664f67fe-0894-4a48-b7cd-f20c14c0ffa0-->
<!--title-->
## Retry command for [Build#61092](https://buildkite.com/redpanda/redpanda/builds/61092)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/cluster_config_test.py::ClusterConfigTest.test_rpk_edit_string
tests/rptest/tests/cluster_config_test.py::ClusterConfigTest.test_valid_settings
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--75d51e47-d733-4f35-9ed8-d8cccf8203c4-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/61092">build#61092</a></summary>


| test_id                                                                                                                                                                                                                                           | test_kind   | job_url                                                                                   | test_status   | passed   |
|:--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.cluster_config_test.ClusterConfigTest.test_rpk_edit_string                                                                                                                                                                           | ducktape    | https://buildkite.com/redpanda/redpanda/builds/61092#0194925b-9ca5-443a-b2c4-aad42e315c51 | FAIL          | 0/20     |
| rptest.tests.cluster_config_test.ClusterConfigTest.test_rpk_edit_string                                                                                                                                                                           | ducktape    | https://buildkite.com/redpanda/redpanda/builds/61092#0194925f-e670-4bbc-ad5a-c395b4eb2b33 | FAIL          | 0/20     |
| rptest.tests.cluster_config_test.ClusterConfigTest.test_rpk_export_import                                                                                                                                                                         | ducktape    | https://buildkite.com/redpanda/redpanda/builds/61092#0194925b-9ca6-4c5b-8f21-036ad6960bc4 | FAIL          | 0/20     |
| rptest.tests.cluster_config_test.ClusterConfigTest.test_rpk_export_import                                                                                                                                                                         | ducktape    | https://buildkite.com/redpanda/redpanda/builds/61092#0194925f-e671-436b-8634-ae1696c82644 | FAIL          | 0/20     |
| rptest.tests.cluster_config_test.ClusterConfigTest.test_valid_settings                                                                                                                                                                            | ducktape    | https://buildkite.com/redpanda/redpanda/builds/61092#0194925b-9ca5-443a-b2c4-aad42e315c51 | FAIL          | 0/20     |
| rptest.tests.cluster_config_test.ClusterConfigTest.test_valid_settings                                                                                                                                                                            | ducktape    | https://buildkite.com/redpanda/redpanda/builds/61092#0194925f-e670-4bbc-ad5a-c395b4eb2b33 | FAIL          | 0/20     |
| rptest.tests.random_node_operations_test.RandomNodeOperationsTest.test_node_operations.enable_failures=True.mixed_versions=True.with_tiered_storage=True.with_iceberg=False.with_chunked_compaction=False.cloud_storage_type=CloudStorageType.ABS | ducktape    | https://buildkite.com/redpanda/redpanda/builds/61092#0194925f-e671-436b-8634-ae1696c82644 | FLAKY         | 1/2      |
| test_compat_rpunit.test_compat_rpunit                                                                                                                                                                                                             | unit        | https://buildkite.com/redpanda/redpanda/builds/61092#01949200-c479-449d-97e7-5c4d0dcdc504 | FAIL          | 0/2      |
| test_compat_rpunit.test_compat_rpunit                                                                                                                                                                                                             | unit        | https://buildkite.com/redpanda/redpanda/builds/61092#01949200-c47a-4030-834d-d35ad6578248 | FAIL          | 0/2      |
</details>
<!--/body-->

nvartolomei committed:
config/tests: add simple yaml configuration round trip test

We have an rpk config export/import test which caught a yaml conversion
error while adding a new config property. Adding a unit test to catch
these earlier and allow for quick iteration cycles.

nvartolomei commented:
/dt

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--1c22fa81-178f-414c-8ecf-6262bb5dd643-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/61146">build#61146</a></summary>


| test_id                                                                                                           | test_kind   | job_url                                                                                   | test_status   | passed   |
|:------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.nodes_decommissioning_test.NodesDecommissioningTest.test_recommissioning_one_of_decommissioned_nodes | ducktape    | https://buildkite.com/redpanda/redpanda/builds/61146#01949747-baf8-4f71-b15d-bc1f543c7f95 | FLAKY         | 1/2      |
</details>
<!--/body-->

dotnwat commented:
gots a few conflicts

nvartolomei committed:
kafka/tests: add iceberg property visibility test

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/61196-->
<!--3ba1588d-faa2-4f7a-8dec-39b4a25bdd3b-->
<!--title-->
## Retry command for [Build#61196](https://buildkite.com/redpanda/redpanda/builds/61196)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/datalake/datalake_dlq_test.py::DatalakeDLQPropertiesTest.test_properties
```
<!--/body-->

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--0f378c9f-f6dd-474b-b615-7ce056edd9cf-->
<!--TEST RESULTS-->
<!--0fe08404-edf4-4c34-8558-e290ec46cfc6-->
<!--TEST RESULTS-->
<!--7712a5ae-9eba-40fb-8bbe-7855803c9257-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/61196">build#61196</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:--------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.archival_test.ArchivalTest.test_single_partition_leadership_transfer.cloud_storage_type=CloudStorageType.ABS | ducktape | https://buildkite.com/redpanda/redpanda/builds/61196#01949fbe-1ca3-457c-b2dd-ffb50fdb0376 | FLAKY | 1/2 |
| rptest.tests.datalake.datalake_dlq_test.DatalakeDLQPropertiesTest.test_properties | ducktape | https://buildkite.com/redpanda/redpanda/builds/61196#01949fb9-31da-46a6-a212-9f17b710ff06 | FAIL | 0/20 |
| rptest.tests.datalake.datalake_dlq_test.DatalakeDLQPropertiesTest.test_properties | ducktape | https://buildkite.com/redpanda/redpanda/builds/61196#01949fbe-1ca3-457c-b2dd-ffb50fdb0376 | FAIL | 0/20 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/61199">build#61199</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:-------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.datalake.partition_movement_test.PartitionMovementTest.test_cross_core_movements.cloud_storage_type=CloudStorageType.S3 | ducktape | https://buildkite.com/redpanda/redpanda/builds/61199#0194a261-3811-47a4-8a0f-6fdecfe6c9c7 | FLAKY | 1/2 |
| rptest.tests.partition_balancer_test.PartitionBalancerTest.test_unavailable_nodes | ducktape | https://buildkite.com/redpanda/redpanda/builds/61199#0194a265-1329-46d5-947d-8678a27dabde | FLAKY | 1/2 |
| rptest.tests.partition_state_api_test.PartitionStateAPItest.test_local_summary | ducktape | https://buildkite.com/redpanda/redpanda/builds/61199#0194a261-3810-4146-a9fa-f8ea0fdf71d0 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/61252">build#61252</a></summary>


| test_id                                                                         | test_kind   | job_url                                                                                   | test_status   | passed   |
|:--------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.scaling_up_test.ScalingUpTest.test_scaling_up_with_recovered_topic | ducktape    | https://buildkite.com/redpanda/redpanda/builds/61252#0194aa36-fd8c-4d4a-83df-e7a7405739f1 | FLAKY         | 1/2      |
</details>
<!--/body-->

ztlpn commented on src/v/cluster/topic_properties.cc:
```diff
@@ -43,7 +43,8 @@ std::ostream& operator<<(std::ostream& o, const topic_properties& properties) {
       "delete_retention_ms: {}, "
       "iceberg_delete: {}, ",
-      "iceberg_partition_spec: {}",
+      "iceberg_partition_spec: {}, ",
```
Hmm I think I messed it up when adding `iceberg_partition_spec`. This should be the single string literal, so no commas after "iceberg_delete" and "iceberg_partition_spec". Can you fix it in this PR?

ztlpn commented on src/v/cluster/types.cc:
```diff
@@ -387,7 +387,8 @@ std::ostream& operator<<(std::ostream& o, const incremental_topic_updates& i) {
       "flush_bytes: {}, iceberg_enabled: {}, leaders_preference: {}, "
       "remote_read: {}, remote_write: {}, iceberg_delete: {}, "
-      "iceberg_partition_spec: {}",
+      "iceberg_partition_spec: {}, ",
```
same problem with the string literal here

ztlpn commented on src/v/config/tests/configuration_test.cc:
```diff
@@ -0,0 +1,39 @@
+
+#include <gtest/gtest.h>
+
+ss::logger lg("config test"); // NOLINT
```
nit: probably doesn't matter much, but logger name without spaces is nicer (to specify in a command line etc.)

nvartolomei committed:
cluster/tests: add smoke test for topic properties format string

nvartolomei committed:
cluster: Add missing cmake dependency

nvartolomei committed:
cluster/tests: add smoke test for incremental topic updates format string

nvartolomei committed:
config/tests: remove space from logger

More user friendly when set from command line.

dotnwat reviewed (commented):
wow adding new topic properties is a lot
